lib/asserts.nix

```diff
@@ -50,4 +57,36 @@ rec {
       lib.generators.toPretty {} xs}, but is: ${
         lib.generators.toPretty {} val}";
 
+  /*
+  Assert that a condition holds for all elements of a list. If it doesn't,
+  the first element violating the condition and its index will be given to
+  construct an error message.
+
+  Type:
+    assertAll :: (a -> Bool) -> [a] -> (Int -> a -> String) -> Bool
```

Comment:
Design decisions:
- Should it be called with named arguments instead? E.g.
  ```nix
  # assertAll :: {
  #   condition :: a -> Bool;
  #   list :: [a];
  #   message :: Int -> a -> String;
  # } -> Bool
  assert assertAll {
    condition = n: n != 2;
    list = [ 1 3 5 ];
    message = i: n: "Index ${toString i}, value ${toString n} failed.";
  };
  ```
  [RFC 135](https://github.com/NixOS/rfcs/pull/135) also goes in that direction
- Should the message have access to not only the first, but _all_ elements that don't satisfy the condition? The module system has such a use case: https://github.com/NixOS/nixpkgs/blob/12053de9e5279dcf9d444a99871db7a2615c27d8/lib/modules.nix#L765-L767
  How should this look like? Maybe
  ```nix
  # assertAll :: {
  #   condition :: a -> Bool;
  #   list :: [a];
  #   message :: [{ index :: Int; value :: a; }] -> String;
  # } -> Bool
  assert assertAll {
    condition = n: n != 2;
    list = [ 1 2 3 ];
    message = failingElements: "Index ${(head failingElements).index}, value ${(head failingElements).value} failed.";
  ```

Reply:
1. Yes, and to make it like RFC 135, it should be

``` nix
assertAll = {success, message}: list: ...
```
2. Sounds good. Possibly with an index attached? Needs more thinking and discussion.